File: sdks/php/src/Types/Request.php
<?php

declare(strict_types=1);

namespace Blok\Nanoservice\Types;

/**
 * Request represents the incoming HTTP request data within a workflow context.
 */
final class Request
{
    public function __construct(
        public mixed $body = null,
        /** @var array<string, string> */
        public array $headers = [],
        /** @var array<string, string> */
        public array $params = [],
        /** @var array<string, string> */
        public array $query = [],
        public string $method = '',
        public string $url = '',
        /** @var array<string, string> */
        public array $cookies = [],
        public string $baseUrl = '',
    ) {}

    /**
     * Get a field from the body as a string.
     */
    public function bodyStr(string $key): ?string
    {
        if (is_array($this->body) && isset($this->body[$key]) && is_string($this->body[$key])) {
            return $this->body[$key];
        }
        return null;
    }

    /**
     * Get a typed value from the body.
     */
    public function bodyGet(string $key, mixed $default = null): mixed
    {
        if (is_array($this->body) && array_key_exists($key, $this->body)) {
            return $this->body[$key];
        }
        return $default;
    }

    /**
     * Create a Request from an associative array.
     */
    public static function fromArray(array $data): self
    {
        return new self(
            body: $data['body'] ?? null,
            headers: $data['headers'] ?? [],
            params: $data['params'] ?? [],
            query: $data['query'] ?? [],
            method: $data['method'] ?? '',
            url: $data['url'] ?? '',
            cookies: $data['cookies'] ?? [],
            baseUrl: $data['baseUrl'] ?? '',
        );
    }

    /**
     * Convert to an associative array for JSON serialization.
     */
    public function toArray(): array
    {
        return [
            'body' => $this->body,
            'headers' => $this->headers,
            'params' => $this->params,
            'query' => $this->query,
            'method' => $this->method,
            'url' => $this->url,
            'cookies' => $this->cookies,
            'baseUrl' => $this->baseUrl,
        ];
    }
}

File: sdks/csharp/src/Blok.Core/Types/Request.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Blok.Core.Types;

/// <summary>
/// Request represents the incoming HTTP request data.
/// </summary>
public class Request
{
    [JsonPropertyName("body")]
    public JsonElement Body { get; set; } = default;

    [JsonPropertyName("headers")]
    public Dictionary<string, string> Headers { get; set; } = new();

    [JsonPropertyName("params")]
    public Dictionary<string, string> Params { get; set; } = new();

    [JsonPropertyName("query")]
    public Dictionary<string, string> Query { get; set; } = new();

    [JsonPropertyName("method")]
    public string Method { get; set; } = string.Empty;

    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("cookies")]
    public Dictionary<string, string> Cookies { get; set; } = new();

    [JsonPropertyName("baseUrl")]
    public string BaseUrl { get; set; } = string.Empty;

    /// <summary>
    /// Deserialize the body to a typed value.
    /// </summary>
    public T? BodyAs<T>()
    {
        if (Body.ValueKind == JsonValueKind.Undefined)
            return default;

        return JsonSerializer.Deserialize<T>(Body.GetRawText());
    }

    /// <summary>
    /// Get a string field from the body by key.
    /// </summary>
    public string? BodyString(string key)
    {
        if (Body.ValueKind == JsonValueKind.Object && Body.TryGetProperty(key, out var prop))
        {
            return prop.ValueKind == JsonValueKind.String ? prop.GetString() : prop.GetRawText();
        }
        return null;
    }
}
